### PR TITLE
Merge back JsonWriterUtf8 and JsonWriterUtf16

### PR DESCRIPTION
**Pull request by ahsonkhan**
Reverting the split that was done in https://github.com/dotnet/corefxlab/pull/2292.

There is only a 1-8% regression due to the `if (_isUtf8)` checks.

It isn't enough to justify splitting up the types.

``` ini

BenchmarkDotNet=v0.10.14.534-nightly, OS=Windows 10.0.16299.431 (1709/FallCreatorsUpdate/Redstone3)
Intel Xeon CPU E5-1620 v2 3.70GHz, 1 CPU, 8 logical and 4 physical cores
Frequency=3604597 Hz, Resolution=277.4235 ns, Timer=TSC
.NET Core SDK=2.2.100-preview1-008958
  [Host]     : .NET Core 2.1.0-rtm-26515-03 (CoreCLR 4.6.26515.03, CoreFX 4.6.26515.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rtm-26515-03 (CoreCLR 4.6.26515.03, CoreFX 4.6.26515.03), 64bit RyuJIT


```
|                         Method | IsUTF8Encoded | Formatted |         Mean |       Error |      StdDev |  Gen 0 | Allocated |
|------------------------------- |-------------- |---------- |-------------:|------------:|------------:|-------:|----------:|
|      **WriterSystemTextJsonBasic** |         **False** |     **False** | **13,529.82 ns** | **159.0390 ns** | **148.7652 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |         False |     False | 25,062.88 ns |  31.7473 ns |  26.5105 ns | 0.0610 |     472 B |
| WriterSystemTextJsonHelloWorld |         False |     False |     85.79 ns |   0.2155 ns |   0.1911 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |         False |     False |    267.93 ns |   0.9935 ns |   0.7757 ns | 0.0591 |     312 B |
|      **WriterSystemTextJsonBasic** |         **False** |      **True** | **15,564.28 ns** |  **48.1039 ns** |  **40.1689 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |         False |      True | 38,421.90 ns | 734.3675 ns | 686.9277 ns | 0.0610 |     640 B |
| WriterSystemTextJsonHelloWorld |         False |      True |     92.68 ns |   0.2296 ns |   0.1660 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |         False |      True |    390.95 ns |   1.2159 ns |   0.8792 ns | 0.0911 |     480 B |
|      **WriterSystemTextJsonBasic** |          **True** |     **False** | **13,743.44 ns** | **265.1843 ns** | **260.4464 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |          True |     False | 24,569.25 ns |  66.1733 ns |  47.8476 ns | 0.0610 |     416 B |
| WriterSystemTextJsonHelloWorld |          True |     False |    139.25 ns |   2.7081 ns |   2.7810 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |          True |     False |    310.22 ns |   1.0948 ns |   0.9142 ns | 0.0486 |     256 B |
|      **WriterSystemTextJsonBasic** |          **True** |      **True** | **15,962.09 ns** | **103.2428 ns** |  **86.2124 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |          True |      True | 37,516.61 ns |  58.4884 ns |  51.8484 ns | 0.0610 |     584 B |
| WriterSystemTextJsonHelloWorld |          True |      True |    152.16 ns |   2.4260 ns |   2.0258 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |          True |      True |    420.69 ns |   5.1807 ns |   4.0448 ns | 0.0806 |     424 B |

### CODE REVIEW

File: src/System.Text.JsonLab/System/Text/Json/JsonWriter.cs [hunk 1]
```diff
@@ -0,0 +1,1539 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab/System/Text/Json/JsonWriter.cs]: Almost none of this code is new. Most of it is just merging the JsonWriterUtf8 and JsonWriterUtf16 files.

File: src/System.Text.JsonLab/System/Text/Json/JsonWriterUtf8.cs [hunk 2]
```diff
@@ -11,9 +11,12 @@ namespace System.Text.JsonLab
 {
     public struct JsonWriterUtf8
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab/System/Text/Json/JsonWriterUtf8.cs]: I am leaving this as JsonWriterUtf8 for now to make reviewing the merge easy. Once this PR is done, I will rename the type back to JsonWriter.

**Comment by benaadams** [comment]
Any reason these take `IBufferWriter<byte>` rather than `TBufferWriter where : TBufferWriter : IBufferWriter<byte>`

**Comment by benaadams** [comment]
Any reason these take `IBufferWriter<byte>` rather than `TBufferWriter where TBufferWriter : IBufferWriter<byte>`?

`IBufferWriter<byte>` can't be used with any `ref struct` writer that has a cached  `Span<byte>` as it would need boxing (which is disallowed)

Releated, in C#7.3 you can now do an extension method for IBufferWriter : class and TBufferWriter :struct https://github.com/dotnet/roslyn/issues/25025 and it will work through the ambiguity (previously it would choke)

**Comment by KrzysztofCwalina** [comment]
@benaadams, ref structs cannot be used as generic arguments, even if the generic type is also a ref struct.

**Comment by benaadams** [comment]
> ref structs cannot be used as generic arguments, even if the generic type is also a ref struct.

Regular struct then :) Whereas `IBufferWriter<byte>` will force boxing

**Comment by ahsonkhan** [comment]
> Any reason these take `IBufferWriter<byte>` rather than `TBufferWriter where TBufferWriter : IBufferWriter<byte>`?

We have plans to support `BufferWriter<T>` which is similarly defined:
https://github.com/dotnet/corefxlab/blob/826034e72ff4a33e8181b230238e3e7e694b1317/src/System.Buffers.ReaderWriter/System/Buffers/Writer/BufferWriterT.cs#L9

**Comment by benaadams** [comment]
> > Regular struct then

> Which type are you proposing be a regular struct?

Explicitly using the interface in the signature always forces interface dispatch via a shared generic method; using a generic with a constraint allows more options; even if its just for a struct wrapper over a class to allow either inlining or devirtualization of a sealed derived type by forcing a non-shared generic method

**Pull request merged**